x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java

```diff
@@ -428,18 +431,101 @@ public void timingStats(String jobId, Consumer<TimingStats> handler, Consumer<Ex
+                                errorHandler.accept(
+                                    new ElasticsearchException(ExceptionsHelper.shardFailuresToErrorMsg(jobId, shardFailures)));
+                            } else if (unavailableShards > 0) {
+                                errorHandler.accept(
+                                    new ElasticsearchException(
+                                        "[" + jobId + "] Search request encountered [" + unavailableShards + "] unavailable shards"));
+                            } else {
+                                SearchHits hits = searchResponse.getHits();
+                                long hitsCount = hits.getHits().length;
+                                if (hitsCount == 0) {
+                                    SearchRequest searchRequest = msearchRequest.requests().get(i);
+                                    LOGGER.debug("Found 0 hits for [{}]", new Object[]{searchRequest.indices()});
+                                } else if (hitsCount > 1) {
+                                    SearchRequest searchRequest = msearchRequest.requests().get(i);
+                                    LOGGER.debug("Found multiple hits for [{}]", new Object[]{searchRequest.indices()});
+                                } else {
+                                    assert hitsCount == 1;
+                                    SearchHit hit = hits.getHits()[0];
+                                    DatafeedTimingStats timingStats = parseSearchHit(hit, DatafeedTimingStats.PARSER, errorHandler);
+                                    timingStatsByJobId.put(jobId, timingStats);
+                                }
+                            }
+                        }
+                    }
+                    handler.accept(timingStatsByJobId);
+                },
+                errorHandler
+            ),
+            client::multiSearch);
+    }
+
+    public void datafeedTimingStats(String jobId, Consumer<DatafeedTimingStats> handler, Consumer<Exception> errorHandler) {
+        String indexName = AnomalyDetectorsIndex.jobResultsAliasedName(jobId);
+        searchSingleResult(
+            jobId,
+            DatafeedTimingStats.TYPE.getPreferredName(),
+            createLatestDatafeedTimingStatsSearch(indexName, jobId),
+            DatafeedTimingStats.PARSER,
+            result -> handler.accept(result.result),
+            errorHandler,
+            () -> new DatafeedTimingStats(jobId));
+    }
+
+    private SearchRequestBuilder createLatestDatafeedTimingStatsSearch(String indexName, String jobId) {
+        return client.prepareSearch(indexName)
+            .setSize(1)
+            .setIndicesOptions(IndicesOptions.lenientExpandOpen())
+            .setQuery(QueryBuilders.idsQuery().addIds(DatafeedTimingStats.documentId(jobId)))
+            .addSort(SortBuilders.fieldSort(DatafeedTimingStats.TOTAL_SEARCH_TIME_MS.getPreferredName()).order(SortOrder.DESC));
```

Comment:
I don't understand how this is `latest`. I thought `TOTAL_SEARCH_TIME_MS` was a summation of the individual search times. 

Ah, I see, the assumption that it is always monotonically increasing. 

Do we account for the data feed being deleted, and a new one being created? Either we should be keeping the previous data feeds search time total, or we should restart the summation back at zero. 

If we want an actual summation of search time for the lifetime of the job, then continuing to increase the stats from a previously deleted datafeed makes sense. 

If we want an accurate search time per data feed, then this search would fail as we could restart the search time statistics.

Reply:
This is an interesting point.  The datafeed timing stats are persisted with the _job's_ ID and are deleted when the _job_ is deleted.  So internally they are associated with the job rather than the datafeed.  But then they're reported via the datafeed stats endpoint, so externally they appear to be associated with the datafeed.

As long as you always create and delete the job and datafeed together in quick succession it works.  But if the datafeed is deleted and recreated independently then it doesn't.  To make that work the lifecycle of the datafeed stats needs to be moved from the job to the datafeed.  But that is also hard because the datafeed does not have an index - only the job has an index.  Even this is complicated as the job results can be spread over multiple indices, so deletion of the timing stats is not deleting a single document in a single index but potentially multiple documents each with the same ID but in different indices.

@przemekwitek please can you have a look at the delete datafeed code and see how hard it would be to do a delete-by-query to delete the datafeed timing stats when a datafeed is deleted?  I think if it's not too hard we should do this, because externally the stats are part of the datafeed.  But if it's hard let me know and we can think again.

Reply:
> I don't understand how this is `latest`. I thought `TOTAL_SEARCH_TIME_MS` was a summation of the individual search times.
> 
> Ah, I see, the assumption that it is always monotonically increasing.

That's correct.
To make it more explicit I could introduce another field, last_modification_timestamp. BTW, this sounds like something *every* document could benefit from. Is such a functionality built-in or does such a field has to be added on per-index basis?

> Do we account for the data feed being deleted, and a new one being created? Either we should be keeping the previous data feeds search time total, or we should restart the summation back at zero.
> 
> If we want an actual summation of search time for the lifetime of the job, then continuing to increase the stats from a previously deleted datafeed makes sense.
> 
> If we want an accurate search time per data feed, then this search would fail as we could restart the search time statistics.

Good point, so far I was treating job-datafeed pair as a unity so I have to revisit that.

Reply:
> This is an interesting point. The datafeed timing stats are persisted with the _job's_ ID and are deleted when the _job_ is deleted. So internally they are associated with the job rather than the datafeed. But then they're reported via the datafeed stats endpoint, so externally they appear to be associated with the datafeed.
> 
> As long as you always create and delete the job and datafeed together in quick succession it works. But if the datafeed is deleted and recreated independently then it doesn't. To make that work the lifecycle of the datafeed stats needs to be moved from the job to the datafeed. But that is also hard because the datafeed does not have an index - only the job has an index.

Correct, I did not want to introduce yet another index, especially that we do not plan to foster job-datafeed divergence.

> Even this is complicated as the job results can be spread over multiple indices, so deletion of the timing stats is not deleting a single document in a single index but potentially multiple documents each with the same ID but in different indices.
> 
> @przemekwitek please can you have a look at the delete datafeed code and see how hard it would be to do a delete-by-query to delete the datafeed timing stats when a datafeed is deleted? I think if it's not too hard we should do this, because externally the stats are part of the datafeed. But if it's hard let me know and we can think again.

I'll do that.

Reply:
This is now done.

PTAL